### PR TITLE
Product Collection - New 'No Results' block with default UI

### LINKED ISSUE #10826: Product Collection: Enhance No Results View in Block
Related GitHub discussion: https://github.com/woocommerce/woocommerce-blocks/discussions/10707

### Summary

- The goal of this task is to improve the "No Results" view in the Product Collection block. The task will involve offering default content, introducing patterns, and examining the visibility toggle options in List View.

---

### Background

- The "No Results" block currently always shows up at the bottom of the Product Collection block, with an empty Paragraph block as the default content. This design can be cumbersome to work with, especially when multiple Collection blocks are present on the same page. We aim to provide a more dynamic, user-friendly, and customizable experience.

---

### Designs

- Figma: zx83mG5ZnDiacPdH6GlO2j-fi-399_96156
- **Current No Results block Screenshot**
image

- **Proposed Changes Screenshot**
image

- **Visibility Toggle Video**

https://github.com/woocommerce/woocommerce-blocks/assets/191598/46adb42d-f2df-4366-925b-c3f1292a1c99

---

### Acceptance Criteria

- **Default Content**: Implement default content that automatically appears in the No Results block.
- **Design Patterns**: Introduce design patterns that offer customization options for merchants and builders.
- **Icon Change**: Update the icon for the block to the Conditional Content icon.
- **Description Update**: Make the description more readable and understandable.
- **Visibility**: Introduce a mechanism to show the No Results block only when the parent Collection block (or one of its children) is selected.
  - **Hidden Indicator in List View**: Investigate if it's possible to add a new "hidden" indicator icon that can toggle the visibility of the No Results block in the List View.

---

### Feedback Loop

- This task will involve a feedback loop with @shaunandrews . After the initial implementation of the features, @shaunandrews will review the changes and may suggest further enhancements or modifications. This is to ensure that the improved No Results block aligns with the overall UX goals.

---

### Additional Notes

- The visibility toggle feature may require contributions to Gutenberg core.
- This improvement can also benefit other blocks like Mini Cart, and Query Loop, which have similar "No Results" views.

CC: @kmanijak

### PR DESCRIPTION
**Pull request by imanish003**
## What

This PR introduces a custom 'No Results' block for the Product Collection block, specifically designed to replace the default core 'query-no-results' block usage. This new block provides the following default view when there are no products for selected filters.

image

Part of #10826

## Why

Developing a new "No Results" block is essential for an improved default user interface. This initiative will lay the groundwork for enhancing the user experience for merchants. Additionally, it will facilitate the creation of extensibility options for third-party developers (3PDs). Furthermore, we plan to introduce a pattern selection feature, enabling merchants to customize the appearance of the "No Results" view. For more info, you may also check Github Discussion #10707.

## Testing Instructions

_Please consider any edge cases this change may have, and also other areas of the product this may impact._

### Case 1: Normal

1. Create a new post & add Product Collection block
2. Verify that there is "No results" block, which UI should look like this:
image

    
3. Save the post & verify on Frontend that "No results" blocks work as expected:
    - **Note:** This block will be only visible if there are no products to show on the frontend. You can achieve this by changing the filters in Inspector control such that there are no products to show for those filters. For example, try setting "Stock Status" filter to only "On backorder".
4. There are two links on frontend:
    - **store's home:** Verify that "store's home" link takes you to the home page of the store.
    - **clearing any filters:** This functionality is tested in Case 2.

### Case 2: Testing `clearing any filters` link

1. Create a new post.
2. Add filters blocks, i.e. Filter by Price, Filter by Stock, Filter by Attribute, Filter by Rating, etc. You can use a pattern called "Product Filters" which contains all the filters.
3. Add Product Collection block & publish the post
4. Go to frontend:
    - Change all filter block values so that there are no products for selected filters and no results block becomes visible. URL will look something like this: `http://store.local/162-2/?filter_stock_status=instock&filter_color=blue%2Cgray&query_type_color=or&min_price=29&max_price=32`
    - Click on "clearing any filters" link & verify that it clears all filter values.
    - *TO BE TESTED BY DEVELOPERS*
        - You can also add some random query variables to URL, for example `&random=value` & URL will look something like this: `http://store.local/162-2/?filter_stock_status=instock&filter_color=blue%2Cgray&query_type_color=or&min_price=29&max_price=32&random=value`
            - Verify that clicking on `clearing any filters` only removes query variables that are related to filters & doesn't remove this random query variable we added.

### Case 3: Adding styles to "No results" block

1. Create a new post & add Product Collection block.
2. Select "No results" block & change styles using Inspector control, i.e.:
    - Text, Background & Link color
    - Typography: Font family, appearance, line height, decoration etc.
3. Verify the changes made to these styles are visible on Editor.
4. Publish the post & verify that changes are visible on Frontend as well.

### Case 4: Adding/removing blocks

1. Create a new post & add Product Collection block.
2. In “No results” block:
    1. Verify that you can remove existing blocks.
    2. Verify that you can add new blocks to it.
    3. Verify that the above changes are reflected correctly on the frontend.

Here is a quick demo:

https://github.com/woocommerce/woocommerce-blocks/assets/16707866/a5907a6f-8131-45f9-9c18-6f693cd02b5c

* [ ] Do not include in the Testing Notes 
* [ ] Should be tested by the development team exclusively 

## Screenshots or screencast

| Before | After |
| ------ | ----- |
|        |       |

## WooCommerce Visibility

Required:

* [X] WooCommerce Core
* [ ] Feature plugin
* [ ] Experimental
* [ ] N/A

## Checklist

Required:

* [X] This PR has either a `[type]` label or a `[skip-changelog]` label.
* [ ] This PR is assigned to a milestone.

Conditional:

* [x] This PR has a UI change and has been cross-browser tested at different viewport sizes on both the frontend and in the editor.
* [x] This PR has a changelog description (if `[skip-changelog]` label is not present).
* [ ] This PR adds/removes a feature flag & I've updated [this doc](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/docs/internal-developers/blocks/feature-flags-and-experimental-interfaces.md).
* [ ] This PR adds/removes an experimental interfaces, and I've updated [this doc](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/docs/internal-developers/blocks/feature-flags-and-experimental-interfaces.md).
* [ ] This PR has been accessibility tested.
* [ ] This PR has had any necessary documentation added/updated.

## Changelog

> Product Collection - New 'No Results' block with default UI

### CODE REVIEW

File: src/BlockTypes/NoResults.php [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+ * NoResults class.
+ */
+class NoResults extends AbstractBlock {
+
+	/**
+	 * Block name.
+	 *
+	 * @var string
+	 */
+	protected $block_name = 'no-results';
```
Comment (kmanijak) [inline src/BlockTypes/NoResults.php]: May be worth adding a prefix suggesting this block belongs to Product Collection. What do you think?

```suggestion
	protected $block_name = 'product-collection-no-results';
```

File: bin/webpack-entries.js [hunk 2]
```diff
@@ -50,6 +50,9 @@ const blocks = {
 	'product-category': {},
 	'product-categories': {},
 	'product-collection': {},
+	'no-results': {
```
Comment (kmanijak) [inline bin/webpack-entries.js]: May be worth adding a prefix suggesting this block belongs to Product Collection. What do you think?

```suggestion
	'product-collection/no-results': {
```

File: src/BlockTypesController.php [hunk 3]
```diff
@@ -239,6 +239,7 @@ protected function get_block_types() {
 			'ProductCategories',
 			'ProductCategory',
 			'ProductCollection',
+			'NoResults',
```
Comment (kmanijak) [inline src/BlockTypesController.php]: May be worth adding a prefix suggesting this block belongs to Product Collection. What do you think?

```suggestion
			'ProductCollectionNoResults',
```

**Review by kmanijak — APPROVED** [review]
I tested the changes and it works nice! Great work!

I'm thinking about prefixing "No Results" block name in couple of places to indicate it's only created for Product Collection block.

Pre-approving anyway!

File: src/BlockTypes/NoResults.php (hunk 1 above)
Comment (imanish003) [reply src/BlockTypes/NoResults.php]: Agreed! Made the changes in commit https://github.com/woocommerce/woocommerce-blocks/pull/11783/commits/9dc01a9b4adce5cde28645de37bc4d791bd8cbcd.

File: bin/webpack-entries.js (hunk 2 above)
Comment (imanish003) [reply bin/webpack-entries.js]: Agreed! Made the changes in commit https://github.com/woocommerce/woocommerce-blocks/pull/11783/commits/9dc01a9b4adce5cde28645de37bc4d791bd8cbcd.

File: src/BlockTypesController.php (hunk 3 above)
Comment (imanish003) [reply src/BlockTypesController.php]: Made the changes in [0a4395a](https://github.com/woocommerce/woocommerce-blocks/pull/11783/commits/0a4395a839aaddc4d4cdfdc0bf3e656f8244f53e)

**Comment by imanish003** [comment]
Hi @kmanijak, I've implemented the change you requested. Could you please review the PR once more before I proceed with merging it into trunk? Thank you 🙌🏻

**Review by xristos3490 — COMMENTED** [review]
Nice work, @imanish003 👌 

I have two comments based on the screenshot below:


1. How about removing the dot from the heading?
2. These inline paragraphs are creating a strange gap between words. Would it be possible to concat everything into a single paragraph block?

**Comment by imanish003** [comment]
Great spot, @xristos3490 – thank you! 🙏🏻
I've implemented the changes in commit a606add. Could you please review it once more?

**Review by xristos3490 — APPROVED** [review]
Looking good! Thanks, @imanish003!

**Comment by kmanijak** [comment]
> Hi @kmanijak, I've implemented the change you requested. Could you please review the PR once more before I proceed with merging it into trunk? Thank you 🙌🏻

Rechecked after recent changes and all looks good!

PS. Great catches @xristos3490! 🙌

**Pull request merged**